### DISCUSSION TITLE
What'd you do? (good thing!)

### DISCUSSION THREAD

**Discussion by brknsoul** [vladmandic/sdnext#987] [General]
When I started using this fork about 3-4 weeks ago, I was getting maybe 1-1.5 seconds per iteration.
Yesterday I was getting 1.5 _iterations per second_, and now with this latest update, it tops out around 1.78-1.79!

(AMD RX 6600 XT 8Gb, Windows 10, `--use-directml --medvram`)

Awesome work, everyone!

EDIT: 'Tis a beautiful thing;
```
[00:19<00:00,  1.27it/s] # (warmup)
[00:13<00:00,  1.79it/s]
[00:14<00:00,  1.78it/s]
[00:14<00:00,  1.71it/s]
[00:14<00:00,  1.76it/s]
```
![image](https://github.com/vladmandic/automatic/assets/1969381/014c5d55-30da-43f3-a957-b195c4669efb)

**Comment by iDeNoh** [comment]
You could easily double that by installing Linux and ROCM, I'm running a 6700xt 12gb and I'm getting 7.6it/s with those settings

**Reply by brknsoul** [reply]
Yeah , I do that, but I think my commit is too far behind and I cbf reinstalling, plus I don't have all my windows-y things in linux ;-)

I'm patient, I'll just wait until ROCm comes to windows. Should be any time in the next millennium now!

**Reply by tripletto** [reply]
Bought an USB SDD, installed POP OS (Linux with out of the box CUDA support and Nvidia drivers), installed stable diffusion web ui and whenever I feel fancy, I boot from my USB SDD for some Linux SD action. Otherwise I just use my Windows as usual.

**Reply by brknsoul** [reply]
I have a linux partition, but it's solely used for generating.

Windows is only about 4-5 times slower than Linux. And it's slowly catching up.

I mean, if I wanted to do some hardcore genning sessions, then I'd boot into linux. But for the moment it's just a fun thing to do now and then.

**Comment by vladmandic** [comment]
using native (`cuda` or `rocm`) is always going to be faster than using in-between like `directml`.  

but on windows, `rocm` doesn't exist (amd claims they will release it "soon")..  
(don't say `windows is slower than linux`, in this case blame amd for not releasing support for windows at all.)  
plus `rocm` compatibility matrix is really narrow - anything a bit older or new than what they support and you're out of luck.  

so `directml` is the best option for a lot of amd users.  

what did i do regarding `directml`? a1111 was supporting it by simply importing it and hoping it would work. here, its actually utilizing correct device settings throughout the code. additional issue is that `directml` does not have `autocast` capabilities on their own, so we've implemented those manually - result is that fp16 can be used instead of fp32 and that is much faster. but side-effect is that fp16 is still quite a bit sensitive and can cause bad quality images in some cases.
